### ISSUE TITLE
Minor issues with CreateSystemImageSet

### ISSUE THREAD

**Issue by pyscripter** [JAM-Software/Virtual-TreeView#1204]
### Issue 1
CreateSystemImageSet has a parameter pControl,  but it is never called with pControl <> Self.

If pControl was ever to be nil  the last statement below would raise an exception.  (there is a check for nil if IsSystemStyle, but not if it is not).

```pascal
    if StyleServices.Enabled then
      if StyleServices.IsSystemStyle then
      begin
        if Assigned(pControl) then
          Theme := OpenThemeData(pControl.Handle, 'BUTTON')
        else
          Theme := OpenThemeData(Application.Handle, 'BUTTON');
        Details := StyleServices.GetElementDetails(tbCheckBoxUncheckedNormal);
        Res := GetThemePartSize(Theme, BM.Canvas.Handle, Details.Part, Details.State, nil, TS_TRUE, lSize) = S_OK;
      end
      else
        Res := StyleServices.GetElementSize(BM.Canvas.Handle, StyleServices.GetElementDetails(tbCheckBoxUncheckedNormal), TElementSize.esActual, lSize {$IF CompilerVersion >= 34}, pControl.CurrentPPI{$IFEND});
```

Since you never call this method with pControl= nil, that will not cause a problem, but I would suggest that, anyway,  you turn CreateSystemImageSet into a private method and remove the parameter pControl.

###  Issue 2
I think that the whole  `if StyleServices.IsSystemStyle` code block is redundant!  StyleServices.GetElementSize works for both the system and custom styles!  To convince yourself look at:
- TCustomStyleServices.GetElementSize
- TUxThemeStyle.DoGetElementSize
- TCustomStyle.DoGetElementSize

So the whole block above can be replaced with:
```pascal
    if StyleServices.Enabled then
      Res := StyleServices.GetElementSize(BM.Canvas.Handle, StyleServices.GetElementDetails(tbCheckBoxUncheckedNormal), TElementSize.esActual, lSize {$IF CompilerVersion >= 34}, pControl.CurrentPPI{$IFEND});
```
Such a simplification can possibly be applied to other parts of the code (have not checked).

### Issue 3
Instead of using CurrentPPI it is faster and more reliable to use FCurrentPPI.  But this would require, that you move the call to CreateSystemImageSet in ChangeScale after the inherited call:

```pascal
  inherited ChangeScale(M, D{$if CompilerVersion >= 31}, isDpiChange{$ifend});
  if (M <> D) then
  begin
    // Scale utility images, #796
    if FCheckImageKind = ckSystemDefault then begin
      FreeAndNil(FCheckImages);
      if HandleAllocated then
        FCheckImages := CreateSystemImageSet(Self);
    end;
    PrepareBitmaps(True, False); // See issue #991
end;
```
Incidentally it is also worth moving UpdateHeaderRect() after the inherited call, although it won't make a difference.  UpdateHeaderRect calls GetBorderDimensions which calls GetSystemMetrics.  TControl.GetSystemMetrics does not depend on FCurrentPPI:

```pascal
function TControl.GetSystemMetrics(nIndex: Integer): Integer;
begin
  Result := GetSystemMetricsForWindow(nIndex, GetAllocatedWindowHandle);
end;

function GetSystemMetricsForWindow(nIndex: Integer; AHandle: HWnd): Integer;
begin
  if CheckPerMonitorV2SupportForWindow(AHandle)
  then
    Result := WinApi.Windows.GetSystemMetricsForDPI(nIndex, GetDPIForWindow(AHandle))
  else
    Result := WinApi.Windows.GetSystemMetrics(nIndex);
end;
```

**Comment by joachimmarder** [comment]
Thanks, issue 1 and 2 are addressed.

> Incidentally it is also worth moving UpdateHeaderRect() after the inherited call

Agreed.

> Instead of using CurrentPPI it is faster ....

I don't think we should do preemptive performance optimization here.

>  and more reliable to use FCurrentPPI

Can you explain why? Given general best practices, one should use the property, not the member variable.

**Comment by pyscripter** [comment]
@joachimmarder 
Thanks for the changes!

> I don't think we should do preemptive performance optimization here.

It is not about optimization.  As explained in #1205 FCurrentPPI is a protected field and CurrentPPI is a property that gets the PPI from the control handle.  The former is more reliable and consistent with the current control scaling state.  So, _if you can_ (meaning if FCurrentPPI has been updated by the inherited ChangeScale), you should use FCurrentPPI instead of CurrentPPI.

On Issue 2:
You can significantly simplify the code by not checking IsSystemStyle and just using StyleServices.GetElementSize,   Please see above.

In other words the following:

```pascal
      if StyleServices.IsSystemStyle then
      begin
        {$if CompilerVersion >= 33}
        if TOSVersion.Check(10) and (TOSVersion.Build >= 15063)  then
          Theme := OpenThemeDataForDPI(Handle, 'BUTTON', CurrentPPI)
        else
        {$ifend}
          Theme := OpenThemeData(Self.Handle, 'BUTTON');
        Details := StyleServices.GetElementDetails(tbCheckBoxUncheckedNormal);
        Res := GetThemePartSize(Theme, BM.Canvas.Handle, Details.Part, Details.State, nil, TS_TRUE, lSize) = S_OK;
      end
      else
```
can just be removed!
